h2o-core/src/main/java/water/Job.java

```diff
@@ -480,4 +481,42 @@ private void update_from_remote( ) {
       }
   }
   @Override public Class<KeyV3.JobKeyV3> makeSchema() { return KeyV3.JobKeyV3.class; }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected void copyOver(Job src) {
+    super.copyOver(src);
+    _barrier = src._barrier;
+  }
+
+  public static Job<?> tryGetDoneJob(Key<Job> key, long timeMillis) {
+    final Value val = DKV.get(key);
+    if (val == null) {
+      throw new IllegalArgumentException("Job is missing");
+    }
+    final Iced<?> ice = val.get();
+    if (!(ice instanceof Job)) {
+      throw new IllegalArgumentException("Must be a Job not a " + ice.getClass());
+    }
+    final Job<?> j = (Job<?>) ice;
+    if (timeMillis > 0 && !j.isStopped()) {
+      j.waitForDone(timeMillis);
+    }
+    return j;
+  }
+
+  void waitForDone(long timeMillis) {
```

Comment:
I like this idea, maybe the AutoML polling logic could also benefit from it (although I think it should be more radically rethought).

Reply:
that is a good point - to my knowledge, AutoML is using an F/J thread for busy waiting

Reply:
quick attempt: https://github.com/h2oai/h2o-3/pull/6093